### ISSUE TITLE
Android app keep crashing when executing `interpreter.run(input, output)`

### ISSUE THREAD

**Issue by pushpendrapratap** [tensorflow/tensorflow#42850]
**System information** `(for ML model creation and training)`
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): `yes`
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `Linux Ubuntu 20.04`
- TensorFlow installed from (source or binary): `binary`
- TensorFlow version (use command below): `(tf.version.GIT_VERSION: v2.3.0-rc2-23-gb36436b087, tf.version.VERSION: 2.3.0)`
- Python version: `3.8.2`

**System information** `(for running android app)`
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): `yes`
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `Linux Ubuntu 16.04`
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: `Samsung: Galaxy J7`
- TensorFlow installed from (source or binary): `binary`
- TensorFlow version (use command below): `org.tensorflow:tensorflow-lite:0.0.0-nightly, org.tensorflow:tensorflow-lite-select-tf-ops:0.0.0-nightly`
- Android SDK version: `minSdkVersion = 21`, `targetSdkVersion = 29`, `compileSdkVersion = 29`
- Java version: `1.8`
- Kotlin version: `1.3.72-release-Studio4.0-5`

**Describe the current behavior**
- Using Python API: tflite model is able to load and correctly give output
- Using Java API (When used in Android app): tflite model is able to load, gives correct i/p and o/p datatype and shape but it crashes the app when `interpreter.run(input, output)` gets executed

**Describe the expected behavior**
- Using Java API (When used in Android app): it should give the same response as in the case of python API instead of crashing the app

**Standalone code to reproduce the issue** `(tflite model creation and inference using python API)`
```
# tflite model creation:
import numpy as np
import tensorflow as tf

input_names = ["my_input"]
output_names = ["my_output"]
tf_model_path = "../models/r2plus1d.pb"

ip_shape = tf.TensorShape([1, 3, 8, 112, 112])

converter = tf.compat.v1.lite.TFLiteConverter.from_frozen_graph(tf_model_path, input_names, output_names)
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
tflite_model = converter.convert()

r2plus1d_tflite_file = "../models/r2plus1d.tflite"
with open(r2plus1d_tflite_file, "wb") as file:
    file.write(tflite_model)

---------------------------------------------------------------------------------------------
# inference using tflite model:
tflite_model_path = "../models/r2plus1d.tflite"

interpreter = tf.lite.Interpreter(model_path=tflite_model_path)
interpreter.allocate_tensors()

input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

input_shape = input_details[0]['shape']
input_data = np.array(np.random.random_sample(input_shape), dtype=np.float32)
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

output_data = interpreter.get_tensor(output_details[0]['index'])
print(output_data)
```

**Code to reproduce the issue** `(Android app code snippet)` most of its logic has been copied from tensorflow/examples/android
```
package com.myproject.tensorflow

import android.content.res.AssetManager
import android.graphics.Bitmap
import com.google.gson.Gson
import com.myproject.extension.AppLogger
import org.tensorflow.lite.DataType
import org.tensorflow.lite.Interpreter
import java.io.FileInputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel

class TensorFlowImageClassifier private constructor() : Classifier {
    private var counter = 0
    private var isRunning: Boolean = false
    private var interpreter: Interpreter? = null
    override fun recognizeImageForObjectDetection(bitmap: Bitmap): Boolean {
        return false
    }

    if(!isRunning) {
        AppLogger.printLog("action isRunning : "+counter++)
        isRunning = true
        AppLogger.printLog(" launch recognised")
        // val byteBuffer = convertBitmapToByteBufferFinal(bitmap)
        val input = Array(1) {Array(3 ) {Array(8) {Array(112) { FloatArray(112) }}}}
        val result = Array(1) { FloatArray(11) }
        try {
            interpreter?.run(byteBuffer, result)
            // interpreter?.run(input, result)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        AppLogger.printLog("action output")
        AppLogger.printLog("action recognised output outer " + Gson().toJson(result))
    }

    override fun close() {
        interpreter!!.close()
        interpreter = null
    }

    @Throws(IOException::class)
    private fun loadModelFile(
        assetManager: AssetManager,
        modelPath: String
    ): MappedByteBuffer {
        val fileDescriptor = assetManager.openFd(modelPath)
        val inputStream =  FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(  FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }

    private fun convertBitmapToByteBufferFinal(bitmap: Bitmap): ByteBuffer {
        val byteBuffer: ByteBuffer = ByteBuffer.allocateDirect(4 * 1 * 3 * 8 * 112 * 112) // byte
        byteBuffer.order(ByteOrder.nativeOrder())
        for (m in 0 until 3) {
            for (i in 0 until 8) {
                for (j in 0..111) {
                    for (k in 0..111) {
                        byteBuffer.putFloat(0.5f)
                    }
                }
            }
        }
        return byteBuffer
    }

    companion object {
        private lateinit var mAassetManager: AssetManager
        private lateinit var mClassifier: TensorFlowImageClassifier
        private lateinit var mModelPath: String
        lateinit var imageShape: IntArray
        var probabilityDataType: DataType? = null
        var inputDataType: DataType? = null
        lateinit var outputShape: IntArray
        var inputBatchSize = 0
        var imageSizeY = 0
        var imageSizeX = 0
        var pixelValue = 0

        @Throws(IOException::class)
        fun create( assetManager: AssetManager, modelPath: String): Classifier {
            val classifier = TensorFlowImageClassifier()
            classifier.interpreter = getInterpreter(classifier, assetManager, modelPath)
            mAassetManager = assetManager
            mModelPath = modelPath
            mClassifier = classifier
            val probabilityTensorIndex = 0
            val imageTensorIndex = 0
            imageShape = classifier.interpreter!!.getInputTensor(imageTensorIndex).shape() // {1, 3, 8, 112, 112}
            pixelValue = imageShape[1]
            inputBatchSize = imageShape[2]
            imageSizeX = imageShape[3]
            imageSizeY = imageShape[4]
            probabilityDataType = classifier.interpreter!!.getOutputTensor(probabilityTensorIndex).dataType() //FLOAT32
            inputDataType = classifier.interpreter!!.getInputTensor(probabilityTensorIndex).dataType() //FLOAT32
            outputShape = classifier.interpreter!!.getOutputTensor(imageTensorIndex).shape() // {1, 11}
            return classifier
        }

        private fun getInterpreter(
            classifier: TensorFlowImageClassifier,
            assetManager: AssetManager,
            modelPath: String
        ): Interpreter {
            return Interpreter(
                classifier.loadModelFile(assetManager, modelPath),
                Interpreter.Options()
            )
        }
    }
}
```

**Other info / logs**
- [R(2+1)D architecture for action recognition in videos](https://arxiv.org/pdf/1905.00561.pdf)
- [tf model (r2plus1d.pb)](https://drive.google.com/file/d/1in71gffCtQf-eYrx7jz00ZC2IDGbBcVA/view?usp=sharing) - size is around 243MB, used it to convert to tflite
- [tflite model (r2plus1d.tflite)](https://drive.google.com/file/d/1-cUsecwb2HOD0YOFdrhX6BBcRgPHcjUF/view?usp=sharing) - size is around 242MB
- [visualize tflite model (r2plus1d.tflite.svg)](https://drive.google.com/file/d/1Rf8eXHbFH9hGdfwoSI6Jd4xy-KWq64oM/view?usp=sharing) - size is around 1MB
- i/p shape: (1, 3, 8, 112, 112) -> (batch_size, channels, num_of_frames, img_height, img_width)
- o/p shape: (1, 11) -> (num_of_classes)

**Comment by abattery** [comment]
Hi, could you capture the related error logs when android app is crashing?

**Comment by pushpendrapratap** [comment]
Hi, @abattery I'm unable to capture any error logs or trace the error. App just crashes when it reaches to `interpreter?.run(byteBuffer, result)`. Also, tried debugger with breakpoints on that line, it neither prints anything after that line nor going into `catch block`.

**Comment by abattery** [comment]
I ran your model with model_benchmark_plus_flex tool. No crashing found during execution.

```
Peak memory footprint (MB): init=13.0977 overall=1989.27
```

It might be related to an out-of-memory issue since your model requires almost 2GB to run. If you can confirm it, how about reducing the overall model size through some techniques like post training quantization?
https://www.tensorflow.org/lite/performance/post_training_quantization

**Comment by ymodak** [comment]
@pushpendrapratap Is this still an issue? Were you able to apply model optimization techniques as suggested above? Thanks!

**Comment by pushpendrapratap** [comment]
Hi, the NN architecture had mostly `Conv3D` layers, which (as I saw last time) can't be optimized by the current API. Eventually, I realized that this model can't be run on mobile devices and so I went with the `client-server` architecture.

I'm closing this issue but if anyone wants to deploy any action recognition model on mobile devices, I have a few suggestions:
1. Make sure all the operators used by your model can be supported on mobile, however, if that is not the case then try to be creative like `Conv3D -> Conv2D + LSTM` or something similar.
2. See if the same objective can be achieved using the combination of simple models (e.g., object detection + pose estimation model), which requires processing a single frame (at a much faster rate) rather than a bunch of frames (as required by 3D CNNs with a much slower rate).
